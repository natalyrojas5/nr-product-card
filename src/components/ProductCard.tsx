import React from 'react';
import styles from '../styles/styles.module.css';

import useProduct from '../hooks/useProduct';
import { createContext, CSSProperties } from 'react';
import {
  InitialValues,
  onChangeArgs,
  Product,
  ProductCardHandles,
  ProductContextProps,
} from '../interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandles) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  product,
  children,
  className = '',
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });
  return (
    <Provider value={{ counter, increaseBy, product, maxCount }}>
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          reset,
          increaseBy,
        })}
      </div>
    </Provider>
  );
};
