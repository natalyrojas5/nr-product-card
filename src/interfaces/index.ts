import { Props as ProductCardProps } from "../components/ProductCard";
import { Props as ProductTitleProps } from "../components/ProductTitle";
import { Props as ProductImageProps } from "../components/ProductImage";
import { Props as ProductBtnsProps } from "../components/ProductBtns";

export interface Product {
  id: string;
  title: string;
  img?: string;
}

export interface ProductContextProps {
  counter: number;
  product: Product;
  maxCount?: number;
  increaseBy: (value: number) => void;
}

export interface ProductCardHOCProps {
  ({ product, children }: ProductCardProps): JSX.Element;
  Title: ({ title }: ProductTitleProps) => JSX.Element;
  Image: ({ img }: ProductImageProps) => JSX.Element;
  Buttons: ({ className }: ProductBtnsProps) => JSX.Element;
}

export interface onChangeArgs {
  product: Product;
  count: number;
}

export interface ProductInCart extends Product {
  count: number;
}

export interface InitialValues {
  count?: number;
  maxCount?: number;
}

export interface ProductCardHandles {
  count: number;
  isMaxCountReached: boolean;
  maxCount?: number;
  product: Product;
  increaseBy: (value: number) => void;
  reset: () => void;
}
