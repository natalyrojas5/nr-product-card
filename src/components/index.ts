import { ProductCard as ProductCardHOC } from "./ProductCard";
import { ProductBtns } from "./ProductBtns";
import { ProductImg } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import { ProductCardHOCProps } from "../interfaces";

export { ProductBtns } from "./ProductBtns";
export { ProductImg } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImg,
  Buttons: ProductBtns,
});
