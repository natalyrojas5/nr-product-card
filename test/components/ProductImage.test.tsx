import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImg, ProductCard } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImg', () => {
  test('Debe de mostrar el componente correctamente con la imagen personalizado', () => {
    const wrapper = renderer.create(<ProductImg img="https://hola.jpg" />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  test('Debe de mostrar el componente con la imagen del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImg />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
