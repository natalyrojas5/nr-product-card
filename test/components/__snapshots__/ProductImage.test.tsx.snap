// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImg Debe de mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard "
>
  <img
    alt="Coffe Mug"
    className="productImg undefined"
    src="./coffe-mug.png"
  />
</div>
`;

exports[`ProductImg Debe de mostrar el componente correctamente con la imagen personalizado 1`] = `
<img
  alt="Coffe Mug"
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;
